### PR TITLE
feat(x/poolmanager): split routes swap message

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #4869

## What is the purpose of the change

Implementing new swap message for split routes.

Each route can be thought of as a separate multi-hop swap.

Splitting swaps across multiple pools for the same token pair can be beneficial for several reasons,
primarily relating to reduced slippage, price impact, and potentially lower fees.

Here's a detailed explanation of these advantages:

- **Reduced slippage**: When a large trade is executed in a single pool, it can be significantly affected if someone else executes a large swap against that pool.

- **Lower price impact**: when executing a large trade in a single pool, the price impact can be substantial, leading to a less favorable exchange rate for the trader.
By splitting the swap across multiple pools, the price impact in each pool is minimized, resulting in a better overall exchange rate.

- **Improved liquidity utilization**: Different pools may have varying levels of liquidity, fees, and price curves. By splitting swaps across multiple pools,
the router can utilize liquidity from various sources, allowing for more efficient execution of trades. This is particularly useful when the liquidity in
a single pool is not sufficient to handle a large trade or when the price curve of one pool becomes less favorable as the trade size increases.

- **Potentially lower fees**: In some cases, splitting swaps across multiple pools may result in lower overall fees. This can happen when different pools
have different fee structures, or when the total fee paid across multiple pools is lower than the fee for executing the entire trade in a single pool with
higher slippage.

Note, that the actual split happens off-chain. The router is only responsible for executing the swaps in the order and quantities of token in provided
by the routes.

## Brief Changelog

- generate message and helpers
- add validate basic checks
- implement message route
- add happy path test

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

## TODO

- [x] Add edge case tests
- [x] Add similar implementation for swap amount in given out
- [x] Clean up
- [x] Add changelog entry

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? TBD
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: proto/osmosis/poolmanager/v1beta1/tx.proto [hunk 1]
```diff
@@ -37,6 +37,27 @@ message MsgSwapExactAmountInResponse {
...
+  string token_in_denom = 3
+      [ (gogoproto.moretags) = "yaml:\"token_in_denom\"" ];
+  string token_out_min_amount = 4 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.moretags) = "yaml:\"token_out_min_amount\"",
+    (gogoproto.nullable) = false
+  ];
+}
+
+message MsgSplitRouteSwapExactAmountInInResponse {
```
Comment (jonator) [inline proto/osmosis/poolmanager/v1beta1/tx.proto:53]:
```suggestion
message MsgSplitRouteSwapExactAmountInResponse {
```

File: x/poolmanager/router.go [hunk 2]
```diff
@@ -94,6 +94,64 @@ func (k Keeper) RouteExactAmountIn(
...
+	tokenOutMinAmount sdk.Int,
+) (sdk.Int, error) {
+	if err := types.ValidateSplitRoutes(routes); err != nil {
+		return sdk.Int{}, err
+	}
+
+	var (
+		// We start the multihop min amount as zero because we want
+		// to perform a price impact protection check on the combination of tokens out
+		/// from all multihop paths.
```
Comment (jonator) [inline x/poolmanager/router.go:125]:
```suggestion
		// from all multihop paths.
```

File: x/poolmanager/types/msgs.go [hunk 3]
```diff
@@ -87,3 +90,56 @@ func (msg MsgSwapExactAmountOut) GetSigners() []sdk.AccAddress {
...
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{sender}
+}
+
+// hasDuplicateDeepEqual returns true if there are duplicates
+// in the slice by performing deep comparison. This is useful
+// for comparing matrices or slices of pointers.
+// Returns fals if there are no deep equal duplicates.
```
Comment (jonator) [inline x/poolmanager/types/msgs.go:136]:
```suggestion
// Returns false if there are no deep equal duplicates.
```

**Review by jonator — APPROVED** [review]
A few nits, but overall looks good from a quick scan

File: proto/osmosis/poolmanager/v1beta1/tx.proto [hunk 4]
```diff
@@ -37,6 +37,27 @@ message MsgSwapExactAmountInResponse {
...
+    (gogoproto.nullable) = false
+  ];
+}
+
+message MsgSplitRouteSwapExactAmountInResponse {
+  string token_out_amount = 1 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.moretags) = "yaml:\"token_out_amount\"",
+    (gogoproto.nullable) = false
+  ];
```
Comment (czarcas7ic) [inline proto/osmosis/poolmanager/v1beta1/tx.proto:58]: Would it at all be useful to return a coin instead of just the int? I know it can probably get pulled from the pool itself, but seems like this would be nice as one less step for integrators. Feel free to resolve if disagree, no strong opinion.

File: proto/osmosis/poolmanager/v1beta1/tx.proto [hunk 5]
```diff
@@ -59,3 +80,24 @@ message MsgSwapExactAmountOutResponse {
...
+}
+
+message MsgSplitRouteSwapExactAmountOutResponse {
+  string token_in_amount = 1 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.moretags) = "yaml:\"token_in_amount\"",
+    (gogoproto.nullable) = false
+  ];
+}
```
Comment (czarcas7ic) [inline proto/osmosis/poolmanager/v1beta1/tx.proto:103]: Same question on this

File: x/poolmanager/router.go [hunk 6]
```diff
@@ -93,6 +99,64 @@ func (k Keeper) RouteExactAmountIn(
...
+// It performs the price impact protection check on the combination of tokens out from all multihop paths. The given tokenOutMinAmount
+// is used for comparison.
+//
+// Returns error if:
+//   - routes are empty
+//   - routes contain duplicate multihop paths
+//   - last token out denom is not the same for all multihop paths in route
+//   - one of the multihop swaps fails for internal reasons
+//   - final token out computed is not positive
+//   - final token out computed is smaller than tokenOutMinAmount
```
Comment (czarcas7ic) [inline x/poolmanager/router.go:115]: 🤖

File: x/poolmanager/router_test.go [hunk 7]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+				},
+				{
+					PoolId:        bazUosmoPoolId,
+					TokenOutDenom: baz,
+				},
+			},
+			TokenInAmount: sdk.NewInt(twentyFiveBaseUnitsAmount.Int64() * 3),
+		}
+
+		priceImpactThreshold = sdk.NewInt(97866545)
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1640]: Can you add a comment explaining where this number comes from? It made sense after I read through the tests, but would be nice to have the context prior to then

File: x/poolmanager/router_test.go [hunk 8]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+			routes:            defaultSingleRouteOneHop,
+			tokenInDenom:      foo,
+			tokenOutMinAmount: sdk.OneInt(),
+
+			expectedTokenOutEstimate: twentyFiveBaseUnitsAmount,
+		},
+		"valid solo route multi hop": {
+			routes:            []types.SwapAmountInSplitRoute{defaultSingleRouteTwoHops},
+			tokenInDenom:      foo,
+			tokenOutMinAmount: sdk.OneInt(),
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1668]: asking for my own learning, what is significant about one here? Can this not be 0?

File: x/poolmanager/router_test.go [hunk 9]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+			},
+			tokenInDenom: foo,
+			// equal to the expected amount
+			// every route individually would fail, but the split route should succeed
+			tokenOutMinAmount: priceImpactThreshold,
+
+			// Note: this value was taken from the actual result
+			// and not manually calculated. This is acceptable
+			// for this test because we are not testing the math
+			// but the routing logic.
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1697]: This note should be where we define this var

File: x/poolmanager/router_test.go [hunk 10]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+			expectedTokenOutEstimate: priceImpactThreshold,
+		},
+
+		"error: price impact protection triggerred": {
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1701]:
```suggestion
		"error: price impact protection triggered": {
```

File: x/poolmanager/router_test.go [hunk 11]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+			routes: []types.SwapAmountInSplitRoute{
+				defaultSingleRouteTwoHops,
+				defaultSingleRouteThreeHops,
+			},
+			tokenInDenom: foo,
+			// one greater than expected amount
+			// every route individually would fail, but the split route should succeed
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1708]: This comment says "but the split route should succeed", I think this wasn't meant to be here

File: x/poolmanager/router_test.go [hunk 12]
```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
...
+			tokenOutMinAmount: priceImpactThreshold.Add(sdk.OneInt()),
+
+			// Note: this values were taken from the actual result
+			// and not manually calculated. This is acceptable
+			// for this test because we are not testing the math
+			// but the routing logic.
```
Comment (czarcas7ic) [inline x/poolmanager/router_test.go:1714]: nit: I think its fine to put this comment where we define the var and remove it from under the tests themselves

**Review by czarcas7ic — APPROVED** [review]
Nice job on this, clear and well tested

File: proto/osmosis/poolmanager/v1beta1/tx.proto (hunk 4 above)
Comment (p0mvn) [reply proto/osmosis/poolmanager/v1beta1/tx.proto:58]: The denom is provided by the client as part of the routes so I don't think it's necessary

File: proto/osmosis/poolmanager/v1beta1/tx.proto (hunk 5 above)
Comment (p0mvn) [reply proto/osmosis/poolmanager/v1beta1/tx.proto:103]: https://github.com/osmosis-labs/osmosis/pull/4886/files#r1163604215

File: x/poolmanager/router_test.go (hunk 8 above)
Comment (p0mvn) [reply x/poolmanager/router_test.go:1668]: It can, just small value that we know well not trigger the price impact protection

File: x/poolmanager/router_test.go (hunk 10 above)
Comment (p0mvn) [reply x/poolmanager/router_test.go:1699]: I knew I spelled it wrong lol

File: x/poolmanager/router_test.go (hunk 11 above)
Comment (p0mvn) [reply x/poolmanager/router_test.go]: Fixed

File: x/poolmanager/router.go [hunk 13]
```diff
@@ -305,6 +369,65 @@ func (k Keeper) RouteExactAmountOut(ctx sdk.Context,
 	return tokenInAmount, nil
 }
 
+// SplitRouteExactAmountOut routes the swap across multiple multihop paths
+// to get the desired token in. This is useful for achieving the most optimal execution. However, note that the responsibility
+// of determining the optimal split is left to the client. This method simply routes the swap across the given routes.
+//
+// It performs the price impact protection check on the combination of tokens in from all multihop paths. The given tokenInMaxAmount
+// is used for comparison.
```
Comment (mattverse) [inline x/poolmanager/router.go:377]: nit: consider adding more docs (doesn't have to be as recommended), took some time for me to understand why this is needed or how this is different from `RouteExactAmountOut` 
```suggestion
// SplitRouteExactAmountOut routes the swap across multiple multihop paths
// to get the desired token in, where each route would be multi-hop routes. This allows batching multihop swaps.
//
// This is useful for achieving the most optimal execution. However, note that the responsibility
// of determining the optimal split is left to the client. This method simply routes the swap across the given routes.
//
// It performs the price impact protection check on the combination of tokens in from all multihop paths. The given tokenInMaxAmount
// is used for comparison.
```

File: x/poolmanager/router_test.go [hunk 14]
```diff
@@ -31,6 +39,56 @@ var (
...
+
+	defaultPoolInitAmount     = sdk.NewInt(10_000_000_000)
+	twentyFiveBaseUnitsAmount = sdk.NewInt(25_000_000)
+
+	fooCoin   = sdk.NewCoin(foo, defaultPoolInitAmount)
+	barCoin   = sdk.NewCoin(bar, defaultPoolInitAmount)
+	bazCoin   = sdk.NewCoin(baz, defaultPoolInitAmount)
+	uosmoCoin = sdk.NewCoin(uosmo, defaultPoolInitAmount)
+
+	// Note: These are iniialized in such a way as it makes
```
Comment (mattverse) [inline x/poolmanager/router_test.go:51]:
```suggestion
	// Note: These are initialized in such a way as it makes
```

**Review by mattverse — APPROVED** [review]
LGTM

File: proto/osmosis/poolmanager/v1beta1/swap_route.proto [hunk 15]
```diff
@@ -16,3 +16,23 @@ message SwapAmountOutRoute {
...
+  string token_in_amount = 2 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.moretags) = "yaml:\"token_in_amount\"",
+    (gogoproto.nullable) = false
+  ];
+}
+
+message SwapAmountOutSplitRoute {
+  repeated SwapAmountOutRoute pools = 1
+      [ (gogoproto.moretags) = "yaml:\"pools\"", (gogoproto.nullable) = false ];
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply proto/osmosis/poolmanager/v1beta1/swap_route.proto:32]: It's incorrect to call them routes since each route is a collection of pools. Also, this is coming from FE requirements

File: x/poolmanager/README.md [hunk 16]
```diff
@@ -292,3 +306,28 @@ Example: for converting `ATOM -> OSMO -> LUNA` using two pools with swap fees `0
 instead `0.15% + 0.1%` fees will be aplied. 
 
 [Multi-Hop](https://github.com/osmosis-labs/osmosis/blob/f26ceb958adaaf31510e17ed88f5eab47e2bac03/x/poolmanager/router.go#L16)
+
+## Route Splitting
+
+Each route can be thought of as a separate multi-hop swap.
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/poolmanager/README.md:312]: From my understanding in discussions, routes do not contain routes. Routes contain pools, and the earlier nomenclature in multihop swaps was wrong

File: x/poolmanager/router.go (hunk 13 above)
Comment (p0mvn) [reply x/poolmanager/router.go]: Expanded! However, I'm not sure if calling this batching is correct because we only allow routes on the that start with the same token in and end with the same token out
